pkg/daemon/controller/controller.go

```diff
@@ -0,0 +1,324 @@
+)
+
+type ServiceController struct {
+	svcMgr   winsvc.Mgr
+	client   client.Client
+	ctx      context.Context
+	nodeName string
+}
+
+// RunController is the entry point of WICD's controller functionality
+func RunController(kubeconfigPath string) error {
+	// TODO: This is currently a barebones controller, which pulls only the desired version's ConfigMap,
+	//       and reconciles the VM to match the ConfigMap's specs. This barebones implementation will not watch the
+	//       ConfigMap or the VM for state changes.
+
+	svcMgr, err := winsvc.NewMgr()
+	if err != nil {
+		return err
+	}
+	config, err := clientcmd.BuildConfigFromFlags("", kubeconfigPath)
+	if err != nil {
+		return errors.Wrap(err, "error using kubeconfig to build config")
+	}
+
+	// TODO: This is a client that reads directly from the server, not a cached client like used in WMCO. This should
+	//       be replaced with the cache client received by mgr.GetClient()
+	clientScheme := runtime.NewScheme()
+	err = clientgoscheme.AddToScheme(clientScheme)
+	if err != nil {
+		return err
+	}
+	clientset, err := client.New(config, client.Options{Scheme: clientScheme})
+	if err != nil {
+		return err
+	}
+
+	// use the client to find the name of the node associated with the VM this is running on
+	var nodes core.NodeList
+	err = clientset.List(context.TODO(), &nodes)
+	if err != nil {
+		return err
+	}
+	addrs, err := localIPv4Addresses()
+	if err != nil {
+		return err
+	}
+	node, err := currentNode(&nodes, addrs)
+	if err != nil {
+		return err
+	}
+
+	desiredVersion, present := node.Annotations["windowsmachineconfig.openshift.io/desired-version"]
+	if !present {
+		return errors.New("node missing desired version annotation")
+	}
+	var cm core.ConfigMap
+	err = clientset.Get(context.TODO(), client.ObjectKey{
+		Namespace: "openshift-windows-machine-config-operator", Name: servicescm.NamePrefix + desiredVersion}, &cm)
+	if err != nil {
+		return err
+	}
+	data, err := servicescm.Parse(cm.Data)
+	if err != nil {
+		return err
+	}
+
+	c := NewServiceController(clientset, svcMgr, node.Name)
+	// TODO: Instead of reconciling services only once, we need to watch managed Windows services for changes and
+	//       reconcile services if a change happens, as well as reconciling if any changes occur to the ConfigMap
+	//       or node object.
+	return c.reconcileServices(data.Services)
+}
+
+// localIPv4Addresses returns a slice of all non-loopback ipv4 addresses
+func localIPv4Addresses() ([]net.IP, error) {
+	var ipv4Addresses []net.IP
+	netIfs, err := net.Interfaces()
+	if err != nil {
+		return nil, err
+	}
+	for _, netInterface := range netIfs {
+		addrs, err := netInterface.Addrs()
+		if err != nil {
+			return nil, err
+		}
+		for _, addr := range addrs {
+			ipAddr, ok := addr.(*net.IPNet)
+			if !ok {
+				continue
+			}
+			ipv4Addr := ipAddr.IP.To4()
+			if ipv4Addr != nil || ipv4Addr.IsLoopback() {
+				continue
+			}
+			ipv4Addresses = append(ipv4Addresses, ipv4Addr)
```

Comment:
Hm thanks for catching this, that should be if ipv4==nil

might need to rethink how things are being split up here, since this mistake would have been caught by unit tests
https://github.com/openshift/windows-machine-config-operator/pull/1061#discussion_r869351787